runtime/src/trace.rs

```diff
@@ -0,0 +1,498 @@
+                        }),
+                        metadata.line().map(|line| {
+                            (
+                                opentelemetry::Key::from(CODE_LINENO),
+                                opentelemetry::logs::AnyValue::from(line),
+                            )
+                        }),
+                        metadata.module_path().map(|path| {
+                            (
+                                opentelemetry::Key::from("code.module_path"),
+                                opentelemetry::logs::AnyValue::from(path.to_string()),
+                            )
+                        }),
+                    ]
+                    .into_iter()
+                    .flatten(),
+                ),
+        );
+
+        record.add_attributes(attributes.0);
+
+        self.logger.emit(record)
+    }
+}
+
+/// Convert a [`log::Level`] to its equivalent [`Severity`]
+#[inline(always)]
+pub(crate) fn log_level_as_severity(level: log::Level) -> Severity {
+    match level {
+        log::Level::Info => Severity::Info,
+        log::Level::Warn => Severity::Warn,
+        log::Level::Debug => Severity::Debug,
+        log::Level::Trace => Severity::Trace,
+        log::Level::Error => Severity::Error,
+    }
+}
+
+// Create a Resource that captures information about the entity for which telemetry is recorded.
+pub fn resource(crate_name: &'static str, package_version: &'static str) -> Resource {
+    Resource::from_schema_url(
+        [
+            Some(KeyValue::new(SERVICE_NAME, crate_name)),
+            Some(KeyValue::new(SERVICE_VERSION, package_version)),
+            Some(KeyValue::new(TELEMETRY_SDK_NAME, "opentelemetry")),
+            Some(KeyValue::new(TELEMETRY_SDK_VERSION, "0.27.1")),
+            Some(KeyValue::new(TELEMETRY_SDK_LANGUAGE, "rust")),
+            std::env::var("SHUTTLE_ENV")
+                .ok()
+                .map(|value| KeyValue::new(DEPLOYMENT_ENVIRONMENT_NAME, value)),
+        ]
+        .into_iter()
+        .flatten(),
+        SCHEMA_URL,
+    )
+}
+
+pub fn init_log_subscriber(endpoint: &Option<String>, resource: Resource) -> LoggerProvider {
+    let mut builder = LoggerProvider::builder().with_resource(resource);
+
+    if let Some(endpoint) = endpoint {
+        let exporter = opentelemetry_otlp::LogExporter::builder()
+            .with_http()
+            .with_endpoint(format!("{endpoint}/v1/logs"))
+            .build()
+            .unwrap();
+
+        builder = builder.with_batch_exporter(exporter, runtime::Tokio);
+    }
+
+    builder.build()
+}
+
+// Construct MeterProvider for MetricsLayer
+pub fn init_meter_provider(endpoint: &Option<String>, resource: Resource) -> SdkMeterProvider {
+    let mut builder = MeterProviderBuilder::default().with_resource(resource);
+
+    if let Some(endpoint) = endpoint {
+        let exporter = opentelemetry_otlp::MetricExporter::builder()
+            .with_temporality(Temporality::default())
```

Comment:
Nit: This is already set by default in ::builder().

Reply:
It's included here because 1) I ripped it off wholesale from the example code in `opentelemetry_otlp`, and 2) so that we know what the buttons and levers we have available / might want to fiddle with.

Would you prefer it be omitted?

Reply:
Not important, feel free to leave in.